transforms/universal/fdedup/python/src/signature_calc_transform.py

```diff
@@ -0,0 +1,483 @@
+        # read the target columns
+        df = df.select(self.contents_column, self.document_id_column)
+
+        # generate minhash values
+        minhashes = df.map_rows(
+            lambda row: mm_min_hash.minhash2_nosalt(
+                *self._generate_word_shingles(row, window_size=self.word_shingle_size)
+            )
+        )
+        # rename columns, cast minhashes to list(uint32)
+        minhashes = minhashes.select(
+            pl.col("column_2").alias(self.document_id_column),
+            pl.col("column_0").cast(pl.List(pl.UInt32)).alias("minhashes"),
+            pl.col("column_1").alias("document_length"),
+        )
+        # store the minhash calculations to send out at the end of execution
+        if self.all_minhashes is None:
+            self.all_minhashes = minhashes
+        else:
+            self.all_minhashes = self.all_minhashes.vstack(minhashes)
+
+        # Calculate band hashes
+        band_hashes_list = self.process_rows_into_bands(
+            minhashes,
+            self.num_bands,
+            self.num_rows,
+        )
+        band_hash_schema = pl.Schema(
+            {
+                "band_hash": pl.UInt64,
+                "band_index": pl.Int32,
+                self.document_id_column: pl.Int64,
+            }
+        )
+        band_hashes = pl.DataFrame(band_hashes_list, schema=band_hash_schema)
+
+        # store the band hash calculations to send out at the end of execution
+        if self.all_band_hashes is None:
+            self.all_band_hashes = band_hashes
+        else:
+            self.all_band_hashes = self.all_band_hashes.vstack(band_hashes)
+
+        if len(self.all_minhashes) > 750000:
+            tables, metadata = self.write_band_signatures()
+        else:
+            tables = []
+            metadata = {}
+        # update metadata stats and return the stats (no tables are returned in transform)
+        return tables, metadata
+
+    def flush(self) -> tuple[list[pa.Table], dict[str, Any]]:
+        """
+        This is supporting method for transformers, that implement buffering of tables, for example coalesce.
+        These transformers can have buffers containing tables that were not written to the output. Flush is
+        the hook for them to return back locally stored tables and their statistics. The majority of transformers
+        should use default implementation.
+        If there is an error, an exception must be raised - exit()ing is not generally allowed when running in Ray.
+        :return: a tuple of a list of 0 or more converted tables and a dictionary of statistics that will be
+        propagated to metadata
+        """
+        self.logger.info(f"Starting flush()")
+        if self.all_band_hashes is not None and self.all_minhashes is not None:
+            tables, metadata = self.write_band_signatures()
+        else:
+            tables = []
+            metadata = {}
+        return tables, metadata
+
+    def write_band_signatures(self):
```

Comment:
I would recommend making all methods other than transform() and flush() as protected or private with _ or __ prefixes.